### ISSUE TITLE
undeclared identifier hidden by other diagnostic related to the assumed name

### ISSUE THREAD

**Issue by kaddkaka** [clangd/clangd#547]
When clangd finds an undeclared identifier which is similar to another identifier that exists, it provides the "undeclared identifier" error and a code action to fix it by "correcting" it to the assumed similar identifier, which is good.

However if the assumed identifier would result in a diagnostic error showing, this error shows up even if the code still has the undeclared identifier, which I think is confusing! (It's more problematic in my LSP client because only the first error diagnostic is shown verbosely).

Current behavior when using identifier `force` which clangd thinks is similar to `free`.

**Example code 1**
The problem here is that an error for `free` is showing even if the code currently tries to reference identifier `force`.
```
    1 #include <stdlib.h>
    2
    3 int main() {
W   4   if (!force) ■■ Address of function 'free' will always evaluate to 'true' (fix available)
    5     return 0;
    6 } 
```

**Example code 2**
The only error showing up here is good, code action to rename `force`->`free`.
```
    1 #include <stdlib.h>
    2
    3 int main() {
E   4   if (!&force) ■ Use of undeclared identifier 'force'; did you mean 'free'? (fix available)    /usr/bin/../lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/stdlib.h:61:12:  note: 'free' declared here                                                                                                                   5     return 0;
    6 }
```

**Example code3**
```
    1 #include <stdlib.h>
    2
    3 int main() {
W   4   if (!free) ■ Address of function 'free' will always evaluate to 'true' (fix available)
    5     return 0;
    6 }
```

** Suggestion **
Example code 1 should only have shown the "undeclared identifier" error. Not the error corresponding to incorrect use of `free`.

**System information**
Output of `clangd --version`: Ubuntu clangd version 10.0.1-++20200928083909+ef32c611aa2-1~exp1~20200928185400.194
Editor/LSP plugin: NVIM v0.5.0-706-g42d13b90f  (with builtin LSP client)
Operating system: Ubuntu

**Comment by HighCommander4** [comment]
This doesn't look like a clangd-specific issue to me -- clang itself issues both diagnostics when run on the first example.

**Comment by kaddkaka** [comment]
> This doesn't look like a clangd-specific issue to me -- clang itself issues both diagnostics when run on the first example.

Should I post this as an issue with clang instead?

**Comment by HighCommander4** [comment]
> Should I post this as an issue with clang instead?

Yep, it's probably a good idea to file this at https://bugs.llvm.org/.

**Comment by hokein** [comment]
yeah, agree that the secondary diagnostic is bogus from user perspective -- because of typo-correction, even the source code is written as `if (!force)`, clang AST treats it as `if (!free)` (the ASTs are the same in example 1 and example 3).

From the offline discussion, an ideal is to transform TypoExpr to RecoveryExpr, which will hopefully suppress secondary diagnostics, e.g. in this case instead of transforming to DeclRefExpr, we transforming it to RecoveryExpr.

I'll try to find some time to investigate it.

**Comment by kaddkaka** [comment]
@hokein hi, did you ever have time to look into this? You seemed to have a solution idea
